### ISSUE TITLE
S3 Remote State Owner confusion

### ISSUE THREAD

**Issue by jbruettcva** [hashicorp/terraform#5136]
Saw an issue last week with version 0.6.9 an 0.6.11 whereby authentication to configure remote state and the ownership configuration of the remote state file in an S3 backend may not be valid.  

### The Setup

**main.tf** file includes the following (with approriate values) :
```
provider "aws" {
  access_key = "xxx"
  secret_key = "xxx"
  region = "us-east-1"
  allowed_account_ids = "###"
}
```
Two users have valid AWS credentials setup in the .aws/credentials file, but one user's credentials aren't valid for the account that owns the S3 bucket.

### The Problem
When the user with the invalid account creds configures/pushes new state it creates the state and sets the S3 object owner to the invalid account credentials.  This effectively locks out valid users from accessing that state.

### Suggested Behavior
- Create the S3 object with the bucket owner set as the owner.  
- Authenticate owner user prior to putting file

**Comment by radeksimko** [comment]
There are two ways you can prevent this lock-out from happening:

### Set the S3 bucket policy to enforce `acl`

https://docs.aws.amazon.com/AmazonS3/latest/dev/acl-overview.html#canned-acl
http://docs.aws.amazon.com/AmazonS3/latest/dev/amazon-s3-policy-keys.html

Basically this will cause the S3 bucket to reject any `Put` requests that don't comply with the condition:

```json
{ 
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "statement1",
      "Effect": "Allow",
      "Action": [
        "s3:PutObject"
      ],
      "Resource": [
        "arn:aws:s3:::examplebucket/*"
      ],
      "Condition": {
        "StringEquals": {
          "s3:x-amz-acl": [
            "bucket-owner-full-control"
          ]
        }
      }
    }
  ]
}
```

### Set `acl` for all remote configs

Now most `terraform remote config` that you have already set and subsequent `apply/destroy` is most likely going to fail, so you need to set the `acl`:

```sh
terraform remote config \
    -backend=s3 \
    -backend-config="bucket=terraform-state-prod" \
    -backend-config="key=network/terraform.tfstate" \
    -backend-config="region=us-east-1" \
    -backend-config="acl=bucket-owner-full-control"
```
https://www.terraform.io/docs/state/remote/s3.html

------

in other words this

> Create the S3 object with the bucket owner set as the owner.

is totally possible today.

@jbruettcva Can you confirm it solves your problem?

**Comment by jbruettcva** [comment]
@radeksimko, while I completely agree that your suggestion would be a great _workaround_,  I don't agree that this is not a bug and rather some kind of enhancement request.  The tool is allowing an invalid user to be set as the owner, and it's the tool that's doing it, not the lack of a bucket policy.  Just my opinion.  Thanks for taking the time to respond in either case, you and @jen20 both.  I'll be putting your suggestion in place this afternoon.

**Comment by radeksimko** [comment]
> I don't agree that this is not a bug and rather some kind of enhancement request

I personally don't see this as a bug because the right behaviour/ACL depends on use-cases/workflows that you have in your team/company.

It depends on _who_ owns the bucket and _who_ is using Terraform to save the state in that bucket and what is the trust between these two parties.

It may be just different teams within the same company, in which case `acl=bucket-owner-full-control` in most cases makes sense, but it may also be two different companies where one provides the S3 bucket and gives other company access to it. The company writing into that bucket may not want the owner to have any visibility into the statefile.

You could argue that the first case (2 teams within the company) is the most common case, but we can't draw conclusions based on that and prepare silent surprises for people w/ the 2nd case (effectively expose data that should not have been exposed).

> The tool is allowing an invalid user to be set as the owner, and it's the tool that's doing it, not the lack of a bucket policy.

I don't think it's _invalid_ user - the owner (by default) is the user who uploaded the file initially. That user is still able to update the state file, correct? If not, then that would be definitely a bug.

That said I'm happy to update the documentation example to include `acl=bucket-owner-full-control` so it's even more obvious that you may need to set this if you want to share the state (and expose it via `terraform_remote_state`).

**Comment by jbruettcva** [comment]
We'll agree to disagree.  Thanks again for the input, this issue aside, the tool is fantastic, keep up the great work!
